### PR TITLE
Fixes for error bars and ticks

### PR DESCRIPTION
**Pull request by tdhock**
Hey guys, I coded some fixes for geom_errorbar and geom_errorbarh, which is the first problem in #161 

I changed the way traces are merged at the end of gg2list, so now it does not matter if the error bars come before or after the other geoms (points, lines).

I also moved the definition of a few constants from corresp_one_one.R to the top of ggplotly.R -- this saves me time during development and debugging since I only have to source("ggplotly.R") instead of both code files (in the correct order). I hope this is not a big deal for you guys.

Tests pass on my machine.

### CODE REVIEW

**Comment by mkcor** [comment]
Thanks, @tdhock -- Before I actually review...

> moved the definition of a few constants from `corresp_one_one.R` to the top of `ggplotly.R`

Nooooooooooooooooo!!! We are trying to modularize this code base.

You shouldn't only `source("ggplotly.R")` as you develop, you need the entire package to test properly that your changes are working. Typically, you should use a `Makefile` like:
```bash
all: build
	R --interactive < test_script.R

build:
	R CMD build /dir/to/your/repo/plotly
	R CMD INSTALL --library=/dir/to/your/wip/lib plotly_version#.tar.gz
```
where `test_script.R` contains, say, a new conversion you're working on, so at the end you can make sure that the whole
```
$ make
```
runs successfully. As you develop, you probably insert some `browser()` calls here and there. So then you should
```
$ make build
```
Make sure your `.Rhistory` is never saved. Make sure your WIP version of "plotly" gets installed at the right place (not the default library for R packages). Your `test_script.R` should typically start with
```r
require(plotly, lib.loc="/dir/to/your/wip/lib")
```
Then launch R and `source("test_script.R")`. Finally, to check that the test suite runs well,
```
$ cd tests
$ R
```
and, in the R console,
```r
require(plotly, lib.loc="/dir/to/your/wip/lib")
source("testthat.R")
```

Cool, now this is documented somewhere.

**Comment by tdhock** [comment]
ok sorry about that, I moved the constants back to the other file.

**Comment by tdhock** [comment]
I have also included a bunch of tests from https://github.com/chriddyp/ggplot2-plotly-cookbook/blob/master/axes.R#L70-L71 which I will eventually make pass.

**Comment by mkcor** [comment]
Test-driven development, yes!

File: R/corresp_one_one.R [hunk 1]
```diff
@@ -81,8 +81,12 @@ coded.lty <- c("22"="dash",
...
 
+# Convert R lty line type codes to plotly "dash" codes.
+lty2dash <- c(numeric.lty, named.lty, coded.lty)
+
 # Convert ggplot2 aes to line parameters.
 aes2line <- c(linetype="dash",
               colour="color",
               size="width",
               direction="shape")
+
```
Comment (mkcor) [inline R/corresp_one_one.R]: Remove trailing white line.

**Comment by chriddyp** [comment]
hey @tdhock - our friend @cpsievert is going to start digging in this week, could you check off which features from this list https://github.com/ropensci/plotly/issues/161 this PR is fixing? if you edit my comment and add the associated PR to each of the items, it might make it easier to backtrack in the future

**Comment by tdhock** [comment]
I checked off the items in #161 which I have already fixed in this PR, so @cpsievert can try to fix any of the un-checked items there.

**Comment by tdhock** [comment]
Please code review and check the test table that I set up today.

http://ropensci.github.io/plotly-test-table/

**Comment by chriddyp** [comment]
that table is sweet! I entered these two lines in my chrome console to load jQuery and increase the width of the images to 700 on my big monitor
```
javascript:if(!window.jQuery||confirm('Overwrite\x20current\x20version?\x20v'+jQuery.fn.jquery))(function(d,s){s=d.createElement('script');s.src='https://ajax.googleapis.com/ajax/libs/jquery/1.8/jquery.js';(d.head||d.documentElement).appendChild(s)})(document);

$('img').css('width', 700)
```

image

**Comment by chriddyp** [comment]
that's funny that you have the border lines in your ggplots, I don't think my ggplots have borders by default

**Comment by chriddyp** [comment]
looks like these 2 got mixed up in the table 
image

**Comment by chriddyp** [comment]
I noticed the "time" and "day" axes labels in your edition, but missing from the ggplot
image

**Comment by chriddyp** [comment]
looks like background color was introduced
image

**Comment by chriddyp** [comment]
I'm assuming the far-left image was out of order 
image

**Comment by chriddyp** [comment]
missing a box here: 
image

**Comment by tdhock** [comment]
there was some bug with the test table I uploaded last night, but now it is fixed.

I also documented the workflow for adding a column to the table here

https://github.com/ropensci/plotly-test-table

**Comment by mkcor** [comment]
@tdhock I love the table!!  It would make more sense to me to have `ggplot2` as the first column, `master` as the middle column, and `feature-branch` as the third column...  Because that's almost a timeline.  I can see how you would want `feature-branch` and `ggplot2` side by side, but...  Anyway, not a big deal.

**Comment by tdhock** [comment]
@cpsievert check out the test table that I put together. 

@mkcor when merging any new branch with master, there are two interesting comparisons to make: (1) do the new branch plots look the same as the master plots? and (2) do the new branch plots look like the ggplots? Thus I propose to keep the test table columns ordered as really_old_master old_master current_master new_branch ggplot  (then the interesting comparisons are side by side in the last three columns).

So can I get a +1 to merge with master? or do you want to do more extensive code review?

**Comment by tdhock** [comment]
actually now that I look at the test table it looks like there are a couple of inconsistencies that I still need to work out...

**Comment by chriddyp** [comment]
hey @tdhock - is it possible to have a test table that has those big images again?

**Comment by tdhock** [comment]
You can zoom in to see the big images by clicking on the test name in the first column, for example

http://ropensci.github.io/plotly-test-table/html/ticks-hide-ticks-lines-labels.html

**Comment by tdhock** [comment]
here is a link to the full size table 

http://ropensci.github.io/plotly-test-table/big.html

**Comment by mkcor** [comment]
Reviewing 'extensively' now...

File: NEWS [hunk 2]
```diff
@@ -1,3 +1,20 @@
+TO DISCUSS
```
Comment (mkcor) [inline NEWS]: Please move your notes out of this standard file.

File: R/ggplotly.R [hunk 3]
```diff
@@ -176,8 +176,16 @@ gg2list <- function(p){
     df <- df[order(df$order),]
     df$order <- NULL
 
-    misc$prestats.data <- merge(built$prestats.data[[i]],
-                                gglayout[, c("PANEL", "plotly.row", "COL")])
+    prestats <- built$prestats.data[[i]]
+    ## scale_reverse multiples x/y data by -1, so here we undo that so
```
Comment (mkcor) [inline R/ggplotly.R]: Please use a single `#` for comments (see https://github.com/ropensci/plotly/wiki/Development-guidelines#coding-style and reference therein https://google-styleguide.googlecode.com/svn/trunk/Rguide.xml#comments).

File: R/ggplotly.R [hunk 4]
```diff
@@ -176,8 +176,16 @@ gg2list <- function(p){
...
+    ## that the actual data can be uploaded to plotly.
+    replace.aes <- intersect(names(prestats), reverse.aes)
+    for(a in replace.aes){
```
Comment (mkcor) [inline R/ggplotly.R]: Spaces before `(` and after `)` (see https://google-styleguide.googlecode.com/svn/trunk/Rguide.xml#spacing).

File: R/ggplotly.R [hunk 5]
```diff
@@ -196,7 +204,8 @@ gg2list <- function(p){
     traces <- layer2traces(L, df, misc)
     
     # Associate error bars with previous traces
-    if (grepl("errorbar", L$geom$objname)) {
+    ##if (grepl("errorbar", L$geom$objname)) { #TDH 28 Jan 2015.
+    if(FALSE){
```
Comment (mkcor) [inline R/ggplotly.R]: Is this a debugging leftover?

File: R/ggplotly.R [hunk 6]
```diff
@@ -319,6 +330,29 @@ gg2list <- function(p){
     scale.i <- which(p$scales$find(xy))
     ax.list$title <- if(length(scale.i)){
       sc <- p$scales$scales[[scale.i]]
+      trace.order.list[[xy]] <- sc$limits
+      trace.name.map[sc$breaks] <- sc$labels
+      if(is.null(sc$breaks)){
```
Comment (mkcor) [inline R/ggplotly.R]: Same comment about spacing, below as well.

File: R/ggplotly.R [hunk 7]
```diff
@@ -613,63 +647,103 @@ gg2list <- function(p){
       layout$legend$bgcolor <- toRGB(s(rect_fill))
   }
   
-  trace.list$kwargs <- list(layout=layout)
-  
-  if (length(trace.list) < 2) {
+  if(length(trace.list) == 0) {
```
Comment (mkcor) [inline R/ggplotly.R]: Same comment about spacing.

File: R/ggplotly.R [hunk 8]
```diff
@@ -613,63 +647,103 @@ gg2list <- function(p){
...
-    for (j in 1:nr) {
-      # Use lapply to be elegant?
-      for (d in colnames(comp)) {
-        try(comp[[d]][j] <- trace.list[[j]][[d]], silent=TRUE)
-        # "names" might be NULL in trace.list
+
+  mode.mat <- matrix(NA, 3, 3)
+  rownames(mode.mat) <- colnames(mode.mat) <- c("markers", "lines", "none")
+  mode.mat["markers", "lines"] <-
+    mode.mat["lines", "markers"] <- "lines+markers"
```
Comment (mkcor) [inline R/ggplotly.R]: Fun! I never do this 'two-assignments-at-once'. I believe it could/should be on one line (it's less than 80 characters in length).

File: R/print.R [hunk 9]
```diff
@@ -0,0 +1,4 @@
+print.trace <- function(x, ...){
```
Comment (mkcor) [inline R/print.R]: What's this file?

File: R/trace_generation.R [hunk 10]
```diff
@@ -41,8 +46,8 @@ layer2traces <- function(l, d, misc) {
   
   ## For non-numeric data on the axes, we should take the values from
   ## the original data.
-  for (axis.name in c("x", "y")) {
-    if (!misc$is.continuous[[axis.name]]) {
+  for (axis.name in c("x", "y")) {    
+    if(!misc$is.continuous[[axis.name]]) {
```
Comment (mkcor) [inline R/trace_generation.R]: No, no.

File: tests/testthat.R [hunk 11]
```diff
@@ -3,3 +3,4 @@ save_outputs <- function(gg, name, ignore_ggplot=FALSE) {
   print(paste("running", name))
 }
 test_check("plotly")
+
```
Comment (mkcor) [inline tests/testthat.R]: Nope.

File: tests/testthat/test-ggplot-area.R [hunk 12]
```diff
@@ -3,7 +3,9 @@ context("Area")
 huron <- data.frame(year=1875:1972, level=as.vector(LakeHuron))
 huron$decade <- plyr::round_any(huron$year, 10, floor)
 
-ar <- ggplot(huron) + geom_area(aes(x=year, y=level))
+ar <- ggplot(huron) +
+  theme_grey()+
```
Comment (mkcor) [inline tests/testthat/test-ggplot-area.R]: I don't get the intention here (might it be that you use and abuse `git commit -a`?). If you change the original plot, I'd expect additional/corresponding tests (`expect_` statements). As a general rule, I believe you want to make an additional plot/test, not edit existing plots (unless they have become obsolete...). I guess you want to leave these two plots alone.

File: tests/testthat/test-ggplot-errorbar-horizontal.R [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  L <- gg2list(g)
+
+  # Expect 2 traces
+  expect_equal(length(L), 3)
+  # Expect scatter plot and its error bars to have the same color
+  expect_identical(L[[1]]$marker$color, L[[1]]$error_x$color)
+  expect_identical(L[[2]]$marker$color, L[[2]]$error_x$color)
+  # Expect given errorbar values
+  expect_equal(L[[1]]$error_x$array, c(0.1, 0.3))
+  expect_equal(L[[1]]$error_x$symmetric, TRUE)
```
Comment (mkcor) [inline tests/testthat/test-ggplot-errorbar-horizontal.R]: Use `expect_true()` directly! ;)

File: tests/testthat/test-ggplot-heatmap.R [hunk 14]
```diff
@@ -4,10 +4,12 @@ wdays <- c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday")
 dtimes <- c("Morning", "Afternoon", "Evening")
 workweek <- matrix(c(1, 20, 30, 20, 1, 60, 30, 60, 1, 50, 80, -10, 1, 30, 20),
              nrow=5, ncol=3, byrow=TRUE,
-             dimnames=list(wdays, dtimes))
+             dimnames=list(day=wdays, time=dtimes))
```
Comment (mkcor) [inline tests/testthat/test-ggplot-heatmap.R]: Nice!

File: tests/testthat/test-ggplot-heatmap.R [hunk 15]
```diff
@@ -4,10 +4,12 @@ wdays <- c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday")
...
 ww <- reshape2::melt(workweek)
+ww$day <- factor(ww$day, wdays)
```
Comment (mkcor) [inline tests/testthat/test-ggplot-heatmap.R]: Did you add this just to make it explicit, and/or to handle some possible subsetting?
Because `ww$day` is already a factor...
These tests are hard-coded so I'm not sure that's necessary.

File: tests/testthat/test-ggplot-theme.R [hunk 16]
```diff
@@ -1,6 +1,8 @@
 context("ggplot themes")
 
-iris.base <- ggplot(iris) + geom_point(aes(Petal.Width, Sepal.Width))
+iris.base <- ggplot(iris) +
+  geom_point(aes(Petal.Width, Sepal.Width))+
```
Comment (mkcor) [inline tests/testthat/test-ggplot-theme.R]: Spacing.

File: tests/testthat/test-mean-error-bars.R [hunk 17]
```diff
@@ -0,0 +1,233 @@
+context("means and error bars")
+
+library(ggplot2)
```
Comment (mkcor) [inline tests/testthat/test-mean-error-bars.R]: What is this doing here?

**Comment by mkcor** [comment]
I wish this PR was split into 10 PRs... Would make the review process faster and safer.

File: R/ggplotly.R (hunk 8 above)
Comment (tdhock) [inline R/ggplotly.R]: it does not fit on one line for me.

File: tests/testthat/test-ggplot-heatmap.R (hunk 15 above)
Comment (tdhock) [inline tests/testthat/test-ggplot-heatmap.R]: this is to make sure the days occur in chronological order (not alphabetical order)

File: tests/testthat/test-ggplot-theme.R (hunk 16 above)
Comment (tdhock) [inline tests/testthat/test-ggplot-theme.R]: can you please clarify?

Comment (mkcor) [inline tests/testthat/test-ggplot-theme.R]: There should be a space before the `+` operator.

File: tests/testthat/test-ggplot-heatmap.R (hunk 15 above)
Comment (mkcor) [inline tests/testthat/test-ggplot-heatmap.R]: Okay, thanks! Good to know.

File: R/ggplotly.R (hunk 8 above)
Comment (mkcor) [inline R/ggplotly.R]: What's your editor's config in this respect?
Anyway, I can't argue, because this complies with our guideline "keep the lines to max 80 characters", I was just curious.

**Comment by tdhock** [comment]
The updated test table shows the most recent version of this branch as toby.fixes.reviewed

http://ropensci.github.io/plotly-test-table/tables/c47b8026a456a0ae382011db8ac4cce074fe5b1d/index.html

everything looks fine to me, ready to merge if you guys are.

**Comment by mkcor** [comment]
I'm happy with the changes so far!  If I find something 'missing', I'll submit a PR for you to review @tdhock ;)
:+1: 
/cc @chriddyp

**Comment by chriddyp** [comment]
@tdhock I noticed this `ticks-flip` was fixed but then not here: 
image

**Comment by mkcor** [comment]
@chriddyp Good catch!
@tdhock Do you append '.bug' when running the image test suite on a feature branch, and then '.reviewed' when you re-run it on that same branch once the code has been reviewed?

**Comment by mkcor** [comment]
I think I'll increment the middle digit in version number, after this PR is merged!
On the contrary, the fix for `ticks-flip` will increment the last digit.

**Comment by tdhock** [comment]
@chriddyp turns out my plotly code was correct for that flipped example,

http://ropensci.github.io/plotly-test-table/tables/c47b8026a456a0ae382011db8ac4cce074fe5b1d/ticks-flip.html

but the plotly-test-table code was not. In particular the plotly filename in the JSON kwargs was the same for different versions of the ggplotly code, so actually the non-flipped image that showed up was made using the code from Carson's branch, and the plotly server did not update that image file right away.

But now for the test table I give the the SHA1 in the plotly filename so I hope the plotly server will respond back with a unique plotly every time. Is that a correct interpretation of the filename argument? 

https://github.com/ropensci/plotly-test-table/blob/6a034a3a65e43c03e76072a75118390f998933ed/index.R#L179

**Comment by mkcor** [comment]
I gave you a :+1: a long time ago because you wouldn't do the changes I would ask for so I thought I'll just do them myself...
Can you please remove files `R/print.R`, `tests/testthat.R`, and `tests/testthat/test-ggplot-area.R`? They have nothing to do in this PR!

**Comment by tdhock** [comment]
Sure I can remove those files, if you tell me how?

**Comment by tdhock** [comment]
sorry I am a bit confused. should I just make the reverse changes and then commit them? or is there some other git command?

**Comment by tdhock** [comment]
And I'm sorry if I did not do all the changes you asked for --- I tried to address all of your comments.

**Comment by mkcor** [comment]
Well, I could have been more proactive myself and pulled the branch a while ago. Let me do it now and I'll share my recommendations.

**Comment by mkcor** [comment]
So, I couldn't revert commits, because these dummy changes occurred in the midst of other stuff, i.e., https://github.com/ropensci/plotly/commit/adcc2ba73791247a981440f55ac053cc23c2c724 or https://github.com/ropensci/plotly/commit/4766545186138517c0fe45a91f0f794d97505cf6 (hence my suspicion against `git commit -a`).

Actually I had pulled your branch before, but never pushed. So I did:
```bash
git checkout toby-fixes
git pull origin toby-fixes
git rm R/print.R 
git commit -m "Remove funny file"
# Looked up id/hash for latest commit on master
git checkout 1d75115f719082 tests/testthat.R
git add tests/testthat.R 
git commit -m "Revert testthat.R file to latest commit on master"
git checkout 1d75115f719082 tests/testthat/test-ggplot-area.R
git add tests/testthat/test-ggplot-area.R 
git commit -m "Revert unrelated test file to latest commit on master"
git push origin toby-fixes
```
Hope you find this inspiring!

**Comment by mkcor** [comment]
+1
